### ISSUE TITLE
Roslyn fails to report nullability issues with spread access in collection expressions

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#85698]
**Version Used**: 
```
Insiders [12210.170]
```

**Steps to Reproduce**:

```
#nullable enable
using System.Collections.Generic;
class MyCollection<T>
{
  public int Length { get; }
}
static class Extensions
{
    public static IEnumerator<T> GetEnumerator<T>(this MyCollection<T>? c) => new List<T>().GetEnumerator();
}
class Program
{
    static void Main()
    {
        MyCollection<int>? x = null;
        object[] y = [..x];
    }
}
```

**Diagnostic Id**:
CS8602 - Dereference of a possibly null reference

**Expected Behavior**:
CS8602 is reported for the `..x` spread element because the compiler unconditionally dreferences it to access the `.Length` property

**Actual Behavior**:
No warnings are reported, the program crashes at runtime with a `NullReferenceException`

**Comment by MihuBot** [comment]
I'm a bot. Here are possible related and/or duplicate issues (I may be wrong):
- https://github.com/dotnet/roslyn/issues/74667
- https://github.com/dotnet/roslyn/issues/72121

**Comment by RikkiGibson** [comment]
The bug looks specific to when extension GetEnumerator accepting a null argument is used. When GetEnumerator parameter is non-nullable, or it is an ordinary member, a warning is reported. [.NET Lab](https://lab.razor.fyi/#fZJRSsNAEIYpWpCVgHiCEV_al_QAaVOkllJpQTBPloLb7ZKsJrOandiE0jfxxWPoCTyGp_EAPkiaaGuL3Zdl5_-Z759h2UuVsctY-zGPbGGOv_YTo9CHq8yQjByLrT_tjg5DKUhpNHZPooyV2GXpR1FCfBLKTdO54j5qQ0qYHZLd0VN5hjzMjNryDRQ-bNa8IJZ8qtD_V7A9bu7yXhY7xSQM83AgMb8sJkJuDAyz1QhNz7XY3GIA98kkVAIUEgwk-hTAHHxJDizW1H4Xk0jGnHTc9FzoSVoVanVouUBBrGeQk08ciy0sZoiTElCguylJNPnqSipAo_HTvHTuYjQ9t0aB2pqhDWJJRzmDgTKU--r2RrwiTxGk_BC_KUr2o1ZTGHKFtXpRL-X8_EEqJLcNKbSWozorl57cSkGjMWTQgpFtp-NSXCzpF3txgteHB2-vz0_vnx_to-pNJa18Aw).

**Comment by sergey-karpan** [comment]
Hi, I'd like to work on this issue.

I've been investigating the nullable-flow behavior around collection-expression spreads. The issue appears to be that nullable analysis models the extension GetEnumerator call, but doesn't account for the compiler-generated Length/Count access that may occur before enumeration.

I'm working on a focused fix in this area with regression coverage and will submit a PR if the approach checks out.

**Comment by CyrusNajmabadi** [comment]
> Hi, I'd like to work on this issue.

Go ahead. Ping me when the pr is ready. Thanks :)

**Comment by RikkiGibson** [comment]
Ideally, the bound node for the spread would have information about the Length/Count property which is being used, and we could use, perhaps, the verbose VisitConversion overload, to check the conversion from the receiver argument to the parameter and report any warnings. You'll have to see how the details look for whether that will work out or not.

And, any precedents on how implicitly used members are checked (such as for GetEnumerator itself), may offer a better solution than directly calling VisitConversion.

**Comment by sergey-karpan** [comment]
> Ideally, the bound node for the spread would have information about the Length/Count property which is being used, and we could use, perhaps, the verbose VisitConversion overload, to check the conversion from the receiver argument to the parameter and report any warnings. You'll have to see how the details look for whether that will work out or not.
> 
> And, any precedents on how implicitly used members are checked (such as for GetEnumerator itself), may offer a better solution than directly calling VisitConversion.

Thanks, that helps. I looked into the existing LengthOrCount information on BoundCollectionExpressionSpreadElement and the nullable VisitConversion path.

One complication I found is that LengthOrCount currently appears to represent an available bound candidate, not necessarily an access that lowering will execute. For example, with [1, 2, 3, ..x], ShouldUseKnownLength() can reject the known-length path due to the temporary limit, so x.Length is not emitted. An unconditional nullable analysis of the bound LengthOrCount would therefore produce a false CS8602.

The extension GetEnumerator receiver already eventually goes through the long-form VisitConversion path, so that seems suitable for the receiver-to-parameter nullability check, but it doesn't by itself indicate whether the Length/Count dereference actually occurs.

Would you prefer that the bound representation carry the selected Length/Count usage (so nullable analysis and lowering share the same decision), rather than having NullableWalker duplicate ShouldUseKnownLength() logic?

**Comment by CyrusNajmabadi** [comment]
I'm not ok with duplication.  I am ok if there is a shared helper that both use to compute this on the fly (so that they get the same response).  That said, storing this on the bound node seems reasonable if we are deciding that we will use this member.  and then both consumption points can unify on that.

**Comment by sergey-karpan** [comment]
> I'm not ok with duplication. I am ok if there is a shared helper that both use to compute this on the fly (so that they get the same response). That said, storing this on the bound node seems reasonable if we are deciding that we will use this member. and then both consumption points can unify on that.

Thanks, Cyrus. I checked both options against the current collection-expression lowering.

The stored-decision approach looks workable. The decision is collection-wide, and the information needed to make it is available during collection-expression conversion, so it can be stored on the bound collection expression and consumed by both nullable analysis and lowering without duplicating the selection logic.

I’ve implemented that direction locally and am working through the targeted lowering/regression tests now. I’ll update the PR/issue once the validation is complete.
